### PR TITLE
DDT-825: isElectronic validation

### PR DESCRIPTION
**Pull request by mp-assco**
Dependent on https://github.com/dcsaorg/DCSA-Event-Core/pull/255

### CODE REVIEW

File: src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java [hunk 1]
```diff
@@ -291,6 +298,18 @@
             "number of originals is required for non electronic shipping instructions.");
       }
     }
+    if (Objects.nonNull(shippingInstructionTO.getIsElectronic())
```
Comment (HenrikHL) [inline src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java:301]: Please add a comment for this validation as is done for all other validation rules

File: src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java [hunk 2]
```diff
@@ -291,6 +298,18 @@
...
+        && shippingInstructionTO.getIsElectronic()) {
+      if (Objects.isNull(shippingInstructionTO.getDocumentParties())
+          || shippingInstructionTO.getDocumentParties().isEmpty()) {
+        validationErrors.add("Document parties is required for electronic shipping instructions.");
```
Comment (HenrikHL) [inline src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java:305]:
```suggestion
        validationErrors.add("A documentParty with partyFunction=EBL is required for electronic shipping instructions.");
```

**Comment by HenrikHL** [comment]
I would have assumed that this PR *only* contains code related to the validation - it seems like multiple other files have been modified?!?

**Comment by mp-assco** [comment]
> I would have assumed that this PR _only_ contains code related to the validation - it seems like multiple other files have been modified?!?

It's just formatting

File: src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java (hunk 1 above)
Comment (mp-assco) [reply src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java:301]: Done

File: src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java [hunk 3]
```diff
@@ -291,6 +298,18 @@
...
+        && shippingInstructionTO.getIsElectronic()) {
+      if (Objects.isNull(shippingInstructionTO.getDocumentParties())
+          || shippingInstructionTO.getDocumentParties().isEmpty()) {
+        validationErrors.add("A documentParty with partyFunction=EBL is required for electronic shipping instructions.");
+      }
+      if (shippingInstructionTO.getDocumentParties().stream()
+          .noneMatch(x -> x.getPartyFunction().equals(PartyFunction.EBL))) {
+        validationErrors.add(
+            "For electronic Bill of Laden (isElectronic=true) - an EBL solution provider (PartyFunction=EBL) needs to be provided as a DocumentParty");
+      }
```
Comment (nt-as) [inline src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java:311]: Should we ensure that:

- There is exactly one EBL platform? (What would one do if the shipper listed two?)
- That the party also has an URL? (Or is that handled elsewhere?)

Comment (HenrikHL) [reply src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java:314]: Good point. You need to check the entire list. If multiple EBL platforms are listed - then send an error:
"Only 1 EBL solution provider can be specified in DocumentParties"

Comment (mp-assco) [reply src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java]: Check has been changed to ensure that exactly one EBL is present in DocumentParty when `isElectronic` is true.

File: src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java [hunk 4]
```diff
@@ -292,6 +299,25 @@
       }
     }
 
+    // Check if documentParties is not empty and that partyFunction on DocumentParty set properly
+    // for a non-electronic shipping instruction
```
Comment (HenrikHL) [inline src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java:303]:
```suggestion
    // for an electronic shipping instruction
```

File: src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java [hunk 5]
```diff
@@ -292,6 +299,25 @@
...
+        validationErrors.add(
+            "A documentParty with partyFunction=EBL is required for electronic shipping instructions.");
+      }
+
+      if (Objects.nonNull(shippingInstructionTO.getDocumentParties())
+          && shippingInstructionTO.getDocumentParties().stream()
+                  .filter(x -> x.getPartyFunction().equals(PartyFunction.EBL))
+                  .count()
+              != 1) {
+        validationErrors.add("Only 1 EBL solution provider can be specified in DocumentParties");
```
Comment (HenrikHL) [inline src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java:317]: If number of DocumentParties with EBL = 0 then the error message is wrong... Please make 2 error messages depending on 0 EBL specified and multiple EBL specified

Comment (mp-assco) [reply src/main/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImpl.java]: Done

File: src/test/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImplTest.java [hunk 6]
```diff
@@ -422,10 +428,228 @@ void testCreateShippingInstructionWithCarrierBookingReferenceOnCargoItem() {
...
+
+                verify(locationService, never()).createLocationByTO(any(), any());
+                verify(documentPartyService, never())
+                    .createDocumentPartiesByShippingInstructionReference(any(), any());
+              })
+          .verifyComplete();
+    }
+
+    @Test
+    @DisplayName("Method should save a shipping instruction with isElectronic and EBL documentParties and return shipping response PENC")
```
Comment (HenrikHL) [inline src/test/java/org/dcsa/ebl/service/impl/ShippingInstructionServiceImplTest.java:504]:
```suggestion
    @DisplayName("Method should save a shipping instruction with isElectronic and EBL documentParties and return shipping response DRFT")
```

**Review by HenrikHL — APPROVED** [review]

**Pull request merged**